catalog/app/components/Preview/loaders/IFrame/LoaderBrowsable.tsx

```diff
@@ -0,0 +1,191 @@
+import * as R from 'ramda'
+import * as React from 'react'
+import * as urql from 'urql'
+import * as Sentry from '@sentry/react'
+
+import cfg from 'constants/config'
+import * as Model from 'model'
+import AsyncResult from 'utils/AsyncResult'
+import log from 'utils/Logging'
+import type * as LogicalKeyResolver from 'utils/LogicalKeyResolver'
+import * as PackageUri from 'utils/PackageUri'
+import assertNever from 'utils/assertNever'
+import type { PackageHandle } from 'utils/packageHandle'
+
+import { PreviewError, PreviewData } from '../../types'
+
+import FileType from '../fileType'
+
+import CREATE_BROWSING_SESSION from './CreateBrowsingSession.generated'
+import DISPOSE_BROWSING_SESSION from './DisposeBrowsingSession.generated'
+import REFRESH_BROWSING_SESSION from './RefreshBrowsingSession.generated'
+
+const SESSION_TTL = 60 * 3
+const REFRESH_INTERVAL = SESSION_TTL * 0.2 * 1000
+
+type SessionId = Model.GQLTypes.BrowsingSession['id']
+
+interface ErrorLike {
+  name: string
+  message: string
+}
+
+function mapPreviewError(retry: () => void, e?: ErrorLike) {
+  switch (e?.name) {
+    case 'BucketNotBrowsable':
+      return PreviewError.Forbidden()
+    case 'BucketNotFound':
+      return PreviewError.DoesNotExist()
+    case 'SessionNotFound':
+      return PreviewError.Expired({ retry })
+    case 'OwnerMismatch':
+      return PreviewError.Forbidden()
+    default:
+      return PreviewError.Unexpected({ retry })
+  }
+}
+
+function useCreateSession() {
+  const [, createSession] = urql.useMutation(CREATE_BROWSING_SESSION)
+  return React.useCallback(
+    async (scope: string, ttl) => {
+      const res = await createSession({ scope, ttl })
+      if (res.error) throw res.error
+      if (!res.data) throw new Error('No data')
+      const r = res.data.browsingSessionCreate
+      switch (r.__typename) {
+        case 'BrowsingSession':
+          return r
+        case 'OperationError':
+          throw r
+        case 'InvalidInput':
+          throw new Error(
+            r.errors
+              .map(({ message, path }) => `{message: ${message}, path: ${path} }`)
+              .join('\n'),
+          )
+        default:
+          assertNever(r)
+      }
+    },
+    [createSession],
+  )
+}
+
+function useRefreshSession() {
+  const [, refreshSession] = urql.useMutation(REFRESH_BROWSING_SESSION)
+  return React.useCallback(
+    async (id: string, ttl: number) => {
+      const res = await refreshSession({ id, ttl })
+      if (res.error) throw res.error
+      if (!res.data) throw new Error('No data')
+      const r = res.data.browsingSessionRefresh
+      switch (r.__typename) {
+        case 'BrowsingSession':
+          return r
+        case 'OperationError':
+          throw r
+        case 'InvalidInput':
```

Comment:
it's not necessary in this PR, but i think it would be useful to have reusable OperationError and InvalidInput exception classes with factory methods to create them from gql objects

Reply:
I filled the task in Asana